### PR TITLE
Added an ability to execute more merges and mutations than threads, added new settings

### PR DESCRIPTION
**Pull request by nikitamikhaylov**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):

- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added an ability to execute more merges and mutations than the number of threads in background pool. Merges and mutations will be executed step by step according to their sizes (lower is more prioritized). The ratio of the number of tasks to threads to execute is controlled by a setting `background_merges_mutations_concurrency_ratio`, 2 by default.

Detailed description / Documentation draft:
Implements #22381

### CODE REVIEW

File: src/Core/Settings.h [hunk 1]
```diff
@@ -80,9 +80,14 @@ class IColumn;
     M(Bool, use_uncompressed_cache, false, "Whether to use the cache of uncompressed blocks.", 0) \
     M(Bool, replace_running_query, false, "Whether the running request should be canceled with the same id as the new one.", 0) \
     M(UInt64, background_buffer_flush_schedule_pool_size, 16, "Number of threads performing background flush for tables with Buffer engine. Only has meaning at server startup.", 0) \
-    M(UInt64, background_pool_size, 16, "Number of threads performing background work for tables (for example, merging in merge tree). Only has meaning at server startup.", 0) \
+    M(UInt64, background_pool_size, 16, "Deprecated???", 0) \
```
Comment (KochetovNicolai) [inline src/Core/Settings.h:83]: Not sure what to do with it.
Maybe set it to zero by default, but in case it is specified in config, use it for `background_merges_pool_size` and `background_mutations_pool_size`?

Comment (nikitamikhaylov) [reply src/Core/Settings.h:83]: Maybe throw some kind of exception?

Comment (KochetovNicolai) [reply src/Core/Settings.h:83]: Maybe ... but it would be more difficult to update server

Comment (KochetovNicolai) [reply src/Core/Settings.h:83]: Also the same is for `number_of_free_entries_in_pool_to_lower_max_size_of_merge`

**Review by KochetovNicolai — COMMENTED** [review]
The change is ok itself, the only question is about old settings

File: src/Core/Settings.h (hunk 1 above)
Comment (nvartolomei) [reply src/Core/Settings.h:83]: Keep it and limit merges + mutations to max of that? This way the server could be upgraded and settings tuned later. Also make merges and mutations pool size = to this setting by default.

We have this setting configured to much higher values on systems with good disks/high cpu count.

As I see it the above proposal is the best for easy upgrades. Otherwise we'd have to somehow configure settings prior to upgrade, but that can't be done as older server wouldn't be able to start anymore due to unknown settings...

tldr; `background_merges_pool_size = background_mutations_pool_size = background_pool_size` and `max merges + max mutations = background_pool_size`

ideally...

File: src/Core/Settings.h [hunk 2]
```diff
@@ -80,9 +80,14 @@ class IColumn;
...
+    M(UInt64, background_merges_count, 32, "Number of merges that could be performed in background. Has to be not less than `background_merges_pool_size`. Only has meaning at server startup.", 0) \
```
Comment (nvartolomei) [inline src/Core/Settings.h:84]: `background_merges_count` -> `background_merges_concurrency`? count means well, counting

File: src/Common/CurrentMetrics.cpp [hunk 3]
```diff
@@ -9,8 +9,10 @@
     M(ReplicatedFetch, "Number of data parts being fetched from replica") \
     M(ReplicatedSend, "Number of data parts being sent to replicas") \
     M(ReplicatedChecks, "Number of data parts checking for consistency") \
-    M(BackgroundPoolTask, "Number of active tasks in BackgroundProcessingPool (merges, mutations, or replication queue bookkeeping)") \
+    M(BackgroundMergesPoolTask, "Number of active merges in background pool") \
+    M(BackgroundMutationsPoolTask, "Number of active mutations in background pool") \
     M(BackgroundFetchesPoolTask, "Number of active tasks in BackgroundFetchesPool") \
+    M(BackgroundCommonPoolTask, "Number of active tasks in BackgroundFetchesPool") \
```
Comment (nvartolomei) [inline src/Common/CurrentMetrics.cpp:15]: typo, this is not fetches pool

File: src/Storages/MergeTree/MergeTreeBackgroundExecutor.cpp [hunk 4]
```diff
@@ -14,12 +19,16 @@ String MergeTreeBackgroundExecutor::toString(Type type)
 {
     switch (type)
```
Comment (nvartolomei) [inline src/Storages/MergeTree/MergeTreeBackgroundExecutor.cpp:20]: nit: magic_enum was merged, `return String(magic_enum::enum_name<Type>(type))`

Comment (nikitamikhaylov) [reply src/Storages/MergeTree/MergeTreeBackgroundExecutor.cpp:20]: Thanks!

File: src/Common/CurrentMetrics.cpp (hunk 3 above)
Comment (nikitamikhaylov) [reply src/Common/CurrentMetrics.cpp:15]: Thanks!

File: src/Core/Settings.h (hunk 2 above)
Comment (nikitamikhaylov) [reply src/Core/Settings.h:84]: I have some problems with naming... `background_merges_concurrency` is a good alternative naming for `background_pool_size`, isn't it? Not for tasks count?

Comment (nvartolomei) [reply src/Core/Settings.h:84]: My suggestion is based on this https://techdifferences.com/difference-between-concurrency-and-parallelism.html

If i read the code correctly, then by `count` you actually mean concurrency.

And, by`background_pool_size` you mean _parallelism_. Technically, pool size is good name.

Comment (nvartolomei) [reply src/Core/Settings.h:84]: I'm wrong about assuming pool_size means parallelism. It is also concurrency, just on a different level of abstraction...

File: src/Core/Settings.h [hunk 5]
```diff
@@ -80,9 +80,14 @@ class IColumn;
     M(Bool, use_uncompressed_cache, false, "Whether to use the cache of uncompressed blocks.", 0) \
     M(Bool, replace_running_query, false, "Whether the running request should be canceled with the same id as the new one.", 0) \
     M(UInt64, background_buffer_flush_schedule_pool_size, 16, "Number of threads performing background flush for tables with Buffer engine. Only has meaning at server startup.", 0) \
-    M(UInt64, background_pool_size, 16, "Number of threads performing background work for tables (for example, merging in merge tree). Only has meaning at server startup.", 0) \
+    M(UInt64, background_pool_size, 16, "Deprecated. Number of merges that could be performed in background. Has to be not less than `background_merges_pool_size`. Only has meaning at server startup.", 0) \
+    M(Float, background_merges_concurrency_ratio, 3, "Ratio between a number of how many merges could be processed and a number threads to process them. Only has meaning at server startup.", 0) \
+    M(UInt64, background_merges_pool_size, 8, "Number of threads performing background merges for tables. Only has meaning at server startup.", 0) \
```
Comment (KochetovNicolai) [inline src/Core/Settings.h:85]: Now I don't like that if we have no mutations, than our pool size is 8 (was 16 before)

**Comment by nikitamikhaylov** [comment]
@Mergifyio update

**Comment by nikitamikhaylov** [comment]
@Mergifyio update

File: src/Storages/MergeTree/MergeTreeBackgroundExecutor.h [hunk 6]
```diff
@@ -15,10 +16,123 @@
...
+    std::atomic_bool is_currently_deleting{false};
+    /// Actually autoreset=false is needed only for unit test
+    /// where multiple threads could remove tasks corresponding to the same storage
+    /// This scenario in not possible in reality.
+    Poco::Event is_done{/*autoreset=*/false};
+
+    /// By default priority queue will have max element at top
+    static bool comparePtrByPriority(const TaskRuntimeDataPtr & lhs, const TaskRuntimeDataPtr & rhs)
+    {
+        return lhs->task->getPriority() > rhs->task->getPriority();
```
Comment (KochetovNicolai) [inline src/Storages/MergeTree/MergeTreeBackgroundExecutor.h:51]: A lot of virtual calls, but maybe ok.
Alternative is to add `size_t priority` field and update it every time task is pushed to queue.

File: src/Storages/MergeTree/MergeTreeBackgroundExecutor.h [hunk 7]
```diff
@@ -15,10 +16,123 @@
...
+    }
+};
+
+class IRuntimeQueue
```
Comment (KochetovNicolai) [inline src/Storages/MergeTree/MergeTreeBackgroundExecutor.h:55]: We actually don't need iface here, cause queue is a template. (I am ok to have it).
maybe we can use concepts here :)

File: src/Storages/MergeTree/MergeTreeBackgroundExecutor.h [hunk 8]
```diff
@@ -15,10 +16,123 @@
...
+public:
+    virtual TaskRuntimeDataPtr pop() = 0;
+    virtual void push(TaskRuntimeDataPtr item) = 0;
+    virtual void remove(StorageID id) = 0;
+    virtual void setCapacity(size_t count) = 0;
+    virtual bool empty() = 0;
+    virtual ~IRuntimeQueue() = default;
+};
+
+class OrdinaryRuntimeQueue : public IRuntimeQueue
```
Comment (KochetovNicolai) [inline src/Storages/MergeTree/MergeTreeBackgroundExecutor.h:66]: add final to get rid of virtual calls

File: src/Storages/MergeTree/MergeTreeBackgroundExecutor.h [hunk 9]
```diff
@@ -15,10 +16,123 @@
...
+    bool empty() override { return queue.empty(); }
+
+private:
+    boost::circular_buffer<TaskRuntimeDataPtr> queue{0};
+};
+
+
+/// Consists of two circular buffers
+/// One for merges, one for mutations
+class MergeMutateRuntimeQueue : public IRuntimeQueue
```
Comment (KochetovNicolai) [inline src/Storages/MergeTree/MergeTreeBackgroundExecutor.h:102]: final

File: src/Storages/MergeTree/MergeTreeBackgroundExecutor.h [hunk 10]
```diff
@@ -15,10 +16,123 @@
...
+
+    bool empty() override { return queue.empty(); }
+
+private:
+    boost::circular_buffer<TaskRuntimeDataPtr> queue{0};
+};
+
+
+/// Consists of two circular buffers
+/// One for merges, one for mutations
```
Comment (KochetovNicolai) [inline src/Storages/MergeTree/MergeTreeBackgroundExecutor.h:101]: Comment is outdated

File: src/Storages/MergeTree/MergeTreeSettings.h [hunk 11]
```diff
@@ -42,7 +42,7 @@ struct Settings;
     M(UInt64, max_replicated_mutations_in_queue, 8, "How many tasks of mutating parts are allowed simultaneously in ReplicatedMergeTree queue.", 0) \
     M(UInt64, max_replicated_merges_with_ttl_in_queue, 1, "How many tasks of merging parts with TTL are allowed simultaneously in ReplicatedMergeTree queue.", 0) \
     M(UInt64, number_of_free_entries_in_pool_to_lower_max_size_of_merge, 8, "When there is less than specified number of free entries in pool (or replicated queue), start to lower maximum size of merge to process (or to put in queue). This is to allow small merges to process - not filling the pool with long running merges.", 0) \
-    M(UInt64, number_of_free_entries_in_pool_to_execute_mutation, 10, "When there is less than specified number of free entries in pool, do not execute part mutations. This is to leave free threads for regular merges and avoid \"Too many parts\"", 0) \
+    M(UInt64, number_of_free_entries_in_pool_to_execute_mutation, 20, "When there is less than specified number of free entries in pool, do not execute part mutations. This is to leave free threads for regular merges and avoid \"Too many parts\"", 0) \
```
Comment (KochetovNicolai) [inline src/Storages/MergeTree/MergeTreeSettings.h:45]: I think it is too strict. I think it should not be less than `background_pool_size` at least. So that all threads can do a mutation if there is no merges.
And probably not less - so that merges can delay all mutations (when factor = 2).

**Review by KochetovNicolai — APPROVED** [review]
Ok generally.

**Comment by nikitamikhaylov** [comment]
@Mergifyio update

**Comment by nikitamikhaylov** [comment]
@Mergifyio update

**Comment by nikitamikhaylov** [comment]
@Mergifyio update

**Comment by ka1bi4** [comment]
Internal documentation ticket: DOCSUP-16475
